src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp

```diff
@@ -0,0 +1,354 @@
+    return GetAttribute(endpointId, Attributes::TestInProgress::Id, Attributes::TestInProgress::Get, testInProgress);
+}
+
+bool SmokeCoAlarmServer::GetHardwareFaultAlert(EndpointId endpointId, bool & hardwareFaultAlert)
+{
+    return GetAttribute(endpointId, Attributes::HardwareFaultAlert::Id, Attributes::HardwareFaultAlert::Get, hardwareFaultAlert);
+}
+
+bool SmokeCoAlarmServer::GetEndOfServiceAlert(EndpointId endpointId, EndOfServiceEnum & endOfServiceAlert)
+{
+    return GetAttribute(endpointId, Attributes::EndOfServiceAlert::Id, Attributes::EndOfServiceAlert::Get, endOfServiceAlert);
+}
+
+bool SmokeCoAlarmServer::GetInterconnectSmokeAlarm(EndpointId endpointId, AlarmStateEnum & interconnectSmokeAlarm)
+{
+    return GetAttribute(endpointId, Attributes::InterconnectSmokeAlarm::Id, Attributes::InterconnectSmokeAlarm::Get,
+                        interconnectSmokeAlarm);
+}
+
+bool SmokeCoAlarmServer::GetInterconnectCOAlarm(EndpointId endpointId, AlarmStateEnum & interconnectCOAlarm)
+{
+    return GetAttribute(endpointId, Attributes::InterconnectCOAlarm::Id, Attributes::InterconnectCOAlarm::Get, interconnectCOAlarm);
+}
+
+bool SmokeCoAlarmServer::GetContaminationState(EndpointId endpointId, ContaminationStateEnum & contaminationState)
+{
+    return GetAttribute(endpointId, Attributes::ContaminationState::Id, Attributes::ContaminationState::Get, contaminationState);
+}
+
+bool SmokeCoAlarmServer::GetSensitivityLevel(EndpointId endpointId, SensitivityEnum & sensitivityLevel)
+{
+    return GetAttribute(endpointId, Attributes::SensitivityLevel::Id, Attributes::SensitivityLevel::Get, sensitivityLevel);
+}
+
+chip::BitFlags<Feature> SmokeCoAlarmServer::GetFeatures(EndpointId endpointId)
+{
+    chip::BitFlags<Feature> featureMap;
+    if (!GetAttribute(endpointId, Attributes::FeatureMap::Id, Attributes::FeatureMap::Get, *featureMap.RawStorage()))
+    {
+        ChipLogError(Zcl, "Unable to get the Smoke CO Alarm feature map: attribute read error");
+        featureMap.ClearAll();
+    }
+    return featureMap;
+}
+
+/**********************************************************
+ * SmokeCoAlarmServer private methods
+ *********************************************************/
+
+bool SmokeCoAlarmServer::HandleRemoteSelfTestRequest(CommandHandler * commandObj, const ConcreteCommandPath & commandPath)
+{
+    ExpressedStateEnum expressedState;
+    bool success = GetExpressedState(commandPath.mEndpointId, expressedState);
+
+    if (success)
+    {
+        // If the value is busy then do nothing
+        if (expressedState == ExpressedStateEnum::kSmokeAlarm || expressedState == ExpressedStateEnum::kCOAlarm ||
+            expressedState == ExpressedStateEnum::kTesting || expressedState == ExpressedStateEnum::kInterconnectSmoke ||
+            expressedState == ExpressedStateEnum::kInterconnectCO)
+        {
+            success = false;
+        }
+        else
+        {
+            success = SetTestInProgress(commandPath.mEndpointId, true);
```

Comment:
Spec says ExpressedState should be set to Testing when test starts and then cleared when test stops.  Is the application responsible for doing that?  If so, that should be documented in the header API documentation.